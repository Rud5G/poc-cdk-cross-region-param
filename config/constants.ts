

export const regionUS = 'us-east-1';
export const regionEU = 'eu-west-1';