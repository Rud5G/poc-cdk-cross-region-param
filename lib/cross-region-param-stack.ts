import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as sqs from 'aws-cdk-lib/aws-sqs';


export interface StackProps extends cdk.StackProps {
  env: cdk.Environment;
  crossRegionReferences?: boolean;
}

export class CrossRegionParamStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: StackProps) {
    super(scope, id, props);

    const queueId = `CrossRegionParamQueue-${props.env.region}`;
    // example resource
    const queue = new sqs.Queue(this, queueId, {
      visibilityTimeout: cdk.Duration.seconds(300)
    });
  }
}
