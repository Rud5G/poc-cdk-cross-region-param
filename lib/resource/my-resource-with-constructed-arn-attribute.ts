import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

/**
 * Some resources build their own Arn attribute by constructing strings
 *
 * This will be because the L1 doesn't expose a `{ Fn::GetAtt: ['Arn'] }`.
 *
 * They won't be able to `exportValue()` it, but it shouldn't crash.
 */
export class MyResourceWithConstructedArnAttribute extends cdk.Resource {
    public readonly arn: string;
    public readonly name: string;

    constructor(scope: Construct, id: string, physicalName?: string) {
        super(scope, id, { physicalName });

        const res = new cdk.CfnResource(this, 'Resource', {
            type: 'My::Resource',
            properties: {
                resourceName: this.physicalName,
            },
        });

        this.name = this.getResourceNameAttribute(res.ref.toString());
        this.arn = this.getResourceArnAttribute(cdk.Stack.of(this).formatArn({
            resource: 'my-resource',
            resourceName: res.ref.toString(),
            service: 'myservice',
        }), {
            resource: 'my-resource',
            resourceName: this.physicalName,
            service: 'myservice',
        });
    }
}