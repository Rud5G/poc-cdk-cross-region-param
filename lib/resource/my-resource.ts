import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

export class MyResource extends cdk.Resource {
    public readonly arn: string;
    public readonly name: string;

    constructor(scope: Construct, id: string, physicalName?: string) {
        super(scope, id, { physicalName });

        const res = new cdk.CfnResource(this, 'Resource', {
            type: 'My::Resource',
            properties: {
                resourceName: this.physicalName,
            },
        });

        this.name = this.getResourceNameAttribute(res.ref.toString());
        this.arn = this.getResourceArnAttribute(res.getAtt('Arn').toString(), {
            region: '',
            account: '',
            resource: 'my-resource',
            resourceName: this.physicalName,
            service: 'myservice',
        });
    }
}
