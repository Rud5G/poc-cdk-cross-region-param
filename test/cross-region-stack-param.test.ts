import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { CrossRegionParamStack } from "../lib/cross-region-param-stack";
import { regionEU, regionUS } from "../config/constants";
import { MyResource } from "../lib/resource/my-resource";


describe("CrossRegionStack param", () => {
    let stackEU: CrossRegionParamStack;
    let stackUS: CrossRegionParamStack;
    let templateEU: Template;
    let templateUS: Template;

    beforeAll(() => {
        const app = new cdk.App();
        stackEU = new CrossRegionParamStack(app, `param-stack-${regionEU}`, {
            env: {
                account: process.env.CDK_DEFAULT_ACCOUNT,
                region: regionEU,
            },
            crossRegionReferences: true,
        });

        stackUS = new CrossRegionParamStack(app, `param-stack-${regionUS}`, {
            env: {
                account: process.env.CDK_DEFAULT_ACCOUNT,
                region: regionUS,
            },
            crossRegionReferences: true,
        });

        const myResource = new MyResource(stackEU, 'MyResource');
        new cdk.CfnOutput(stackUS, 'Output', {
            value: myResource.name,
        });

        templateEU = Template.fromStack(stackEU);
        templateUS = Template.fromStack(stackUS);
        // const assembly = app.synth();
        // templateEU = assembly.getStackByName(stackEU.stackName).template;
        // templateUS = assembly.getStackByName(stackUS.stackName).template;
    });

    test("Expect match snapshot for templateEU", () => {
        expect(templateEU.toJSON()).toMatchSnapshot('templateEU');
    });

    test("Expect match snapshot for templateUS", () => {
        expect(templateUS.toJSON()).toMatchSnapshot('templateUS');
    });
});
