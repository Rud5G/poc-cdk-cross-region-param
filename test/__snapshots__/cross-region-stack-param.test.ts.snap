// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CrossRegionStack param Expect match snapshot for templateEU: templateEU 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CrossRegionParamQueueeuwest1F62C52E2": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "VisibilityTimeout": 300,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "CustomCrossRegionExportWriterCustomResourceProviderHandlerD8786E8A": {
      "DependsOn": [
        "CustomCrossRegionExportWriterCustomResourceProviderRoleC951B1E1",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-eu-west-1",
          },
          "S3Key": "8a17fb3ceee821f797100bf65fa3047c164a08c03aa5a07fbcff9a676fb73b97.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomCrossRegionExportWriterCustomResourceProviderRoleC951B1E1",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomCrossRegionExportWriterCustomResourceProviderRoleC951B1E1": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "ssm:DeleteParameters",
                    "ssm:ListTagsForResource",
                    "ssm:GetParameters",
                    "ssm:PutParameter",
                  ],
                  "Effect": "Allow",
                  "Resource": [
                    {
                      "Fn::Join": [
                        "",
                        [
                          "arn:",
                          {
                            "Ref": "AWS::Partition",
                          },
                          ":ssm:us-east-1:",
                          {
                            "Ref": "AWS::AccountId",
                          },
                          ":parameter/cdk/exports/*",
                        ],
                      ],
                    },
                  ],
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ExportsWriteruseast10F67B507DDE2E818": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomCrossRegionExportWriterCustomResourceProviderHandlerD8786E8A",
            "Arn",
          ],
        },
        "WriterProps": {
          "exports": {
            "/cdk/exports/param-stack-us-east-1/paramstackeuwest1RefMyResource6073B41F0F3C838C": {
              "Ref": "MyResource6073B41F",
            },
          },
          "region": "us-east-1",
        },
      },
      "Type": "Custom::CrossRegionExportWriter",
      "UpdateReplacePolicy": "Delete",
    },
    "MyResource6073B41F": {
      "Type": "My::Resource",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`CrossRegionStack param Expect match snapshot for templateUS: templateUS 1`] = `
{
  "Outputs": {
    "Output": {
      "Value": {
        "Fn::GetAtt": [
          "ExportsReader8B249524",
          "/cdk/exports/param-stack-us-east-1/paramstackeuwest1RefMyResource6073B41F0F3C838C",
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CrossRegionParamQueueuseast15652EED3": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "VisibilityTimeout": 300,
      },
      "Type": "AWS::SQS::Queue",
      "UpdateReplacePolicy": "Delete",
    },
    "CustomCrossRegionExportReaderCustomResourceProviderHandler46647B68": {
      "DependsOn": [
        "CustomCrossRegionExportReaderCustomResourceProviderRole10531BBD",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-us-east-1",
          },
          "S3Key": "d41c8e6342cd078b5ea5aec11522bdb605eae00f4bb98a3fb0b44c827e9b5ca9.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomCrossRegionExportReaderCustomResourceProviderRole10531BBD",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomCrossRegionExportReaderCustomResourceProviderRole10531BBD": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": [
          {
            "PolicyDocument": {
              "Statement": [
                {
                  "Action": [
                    "ssm:AddTagsToResource",
                    "ssm:RemoveTagsFromResource",
                    "ssm:GetParameters",
                  ],
                  "Effect": "Allow",
                  "Resource": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":ssm:us-east-1:",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":parameter/cdk/exports/param-stack-us-east-1/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "ExportsReader8B249524": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "ReaderProps": {
          "imports": {
            "/cdk/exports/param-stack-us-east-1/paramstackeuwest1RefMyResource6073B41F0F3C838C": "{{resolve:ssm:/cdk/exports/param-stack-us-east-1/paramstackeuwest1RefMyResource6073B41F0F3C838C}}",
          },
          "prefix": "param-stack-us-east-1",
          "region": "us-east-1",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomCrossRegionExportReaderCustomResourceProviderHandler46647B68",
            "Arn",
          ],
        },
      },
      "Type": "Custom::CrossRegionExportReader",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
